rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/Walkthrough.rsc

```diff
@@ -0,0 +1,290 @@
+// # Walkthrough
+//
+// This module consists of a walkthrough to explain the main ideas behind the
+// conversion algorithm, from Rascal grammars to TextMate grammars.
+//
+// The walkthrough is split into five parts. The initial part explains basic
+// conversion. The subsequent four parts present complications and demonstrate
+// extensions of the conversion algorithm to address them.
+//
+// The toy language considered, is a simple data language consisting of:
+//   - base: maps, numbers, strings;
+//   - extension 1: layout (including comments);
+//   - extension 2: regular expressions;
+//   - extension 3: locations;
+//   - extension 4: booleans.
+//
+// Working familiarity with TextMate grammar is assumed. To summarize:
+//
+//   - Syntax:
+//       - Each TextMate grammar consists of a list of TextMate rules (ordered).
+//       - Each TextMate rule is either a *match pattern* (consisting of one
+//         regular expression) or a *begin/end pattern* (consisting of two
+//         regular expressions and a list of nested TextMate rules).
+//
+//   - Semantics: A tokenization engine reads a document (line by line, top to
+//     bottom, left to right), while iteratively trying to apply TextMate rules
+//     by matching text against the regular expressions.
+//
+// Further reading:
+//   - https://macromates.com/manual/en/language_grammars
+//   - https://www.apeth.com/nonblog/stories/textmatebundle.html
+//   - https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide
+
+module lang::textmate::conversiontests::Walkthrough
+
+import Grammar;
+import ParseTree;
+import lang::textmate::Conversion;
+import lang::textmate::ConversionTests;
+
+import lang::rascal::grammar::analyze::Delimiters;
+
+// ## Basics
+//
+// The base fragment of the grammar looks as follows:
+
+lexical Alnum = [a-z A-Z 0-9];
+lexical Digit = [0-9];
+lexical Blank = [\ \t];
+lexical Space = [\ \t\n];
+
+start syntax Value
+    = Map
+    | Number
+    | String
+    | RegExp
+    | Location
+    | Boolean;
+
+syntax Map = "{" {(Key ":" Value) ","}* "}";
+
+lexical Key    = Alnum+ !>> [a-z A-Z 0-9];
+lexical Number = @category="constant.numeric" Digit+ !>> [0-9];
+lexical String = @category="string.quoted.double" "\"" Alnum* "\"";
+
+// Basically, the conversion algorithm converts each Rascal non-terminal that is
+// *suitable for conversion* to a TextMate rule. For instance, `Number` is
+// converted to the following match pattern (in JSON):
+//
+// ```
+// {
+//   "match": "([\\u0030-\\u0039]+?(?![\\u0030-\\u0039]))",
+//   "name": "prod(lex(\"Number\"),[conditional(iter(lex(\"Digit\")),{\\not-follow(\\char-class([range(48,57)]))})],{tag(\"category\"(\"constant.numeric\"))})", 
```

Comment:
nit: we might have to do something about these names, as they get everywhere in VS Code, and smaller names might be nice for the GC of node?

Reply:
True, in the rascal grammar generator there is a similar scheme for encoding names, maybe we can use the same code for it? But yes, it helps for debugging.

Reply:
I'll look into it reusing that